src/net/socket.rs

```diff
@@ -26,51 +31,72 @@ impl SocketSystem {
    }
 
    /// Resolves the socket addresses the given hostname could refer to.
-   async fn resolve_address_with_default_port(hostname: &str) -> anyhow::Result<Vec<SocketAddr>> {
-      if let Ok(addresses) = lookup_host(hostname).await {
-         Ok(addresses.collect())
-      } else {
-         Ok(lookup_host((hostname, relay::DEFAULT_PORT)).await?.collect())
+   async fn resolve_address_with_default_port(hostname: &str) -> anyhow::Result<url::Url> {
+      let mut url = url::Url::parse(&format!("ws://{}", hostname))?;
```

Comment:
I think we should keep the `ws://` protocol inside the textbox in the lobby, prepending it via simple string concatenation feels very fragile and insecure. Also, having it input by the user allows us to use `wss://`.

Reply:
The only problem is that users probably don't know what `wss://` or `ws://` means and what the difference is between them. In any game, just give the address to the server, as it is now with TCP, and you can connect, which is handy. It's the same in the browser. You don't type `https://example.com`, but simply `example.com` or sometimes `www.example.com`, and the browser selects the protocol itself (with priority on encrypted connection). You could give `wss://` by default in the textbox, but users can remove that, and they won't know what's going on.

By the way, the concatenation here should be safe, because the string is later passed to `Url::parse`, which checks if the address is correct (and will throw an error if the user enters `ws://file://`, for example). After all, it's a whole library to work with url.

Unless you mean that `ws://` is dangerous, which I agree with in general.

Reply:
It's not like users understand what the `http://` in front of website names means, either. Most likely they'll just copy-paste server URLs from somewhere else. If they do not input a scheme we can fall back to `wss://` or `ws://` automatically, in that order of priority.

Reply:
This is what I wanted to do, but for some reason "domain:port" is treated as a scheme. [Here's an example.](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=c5e4e806765c7739e9807422c41737dd).

Reply:
Well, it makes sense from the standpoint of how a URL is constructed.

```ebnf
URI = scheme ":" ["//" authority] path ["?" query] ["#" fragment]
```
<https://en.wikipedia.org/wiki/URL#Syntax>

Then I suppose we can prepend wss:// [in case a user forgets](https://en.wikipedia.org/wiki/DWIM), but if the scheme is already present we should just keep it as is. Though I really do not feel comfortable prepending the scheme like that, I feel like this is bound to have some sort of security vulnerability.

Reply:
It would be best to look to see if crate url has anything that could help with this. I might have missed something.